File: dalesbred/src/test/java/org/dalesbred/DatabasePropertyPathBindingTest.java
/*
 * Copyright (c) 2015 Evident Solutions Oy
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package org.dalesbred;

import org.junit.Rule;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class DatabasePropertyPathBindingTest {

    private final Database db = TestDatabaseProvider.createInMemoryHSQLDatabase();

    @Rule
    public final TransactionalTestsRule rule = new TransactionalTestsRule(db);

    @Test
    public void bindingToNestedPaths() {
        ResultClass result = db.findUnique(ResultClass.class,
                "select 'AAA' as \"nestedField.foo\"," +
                "   'BBB' as \"nestedGetter.foo\"" +
                " from (VALUES (0))");

        assertThat(result.nestedField.foo, is("AAA"));
        assertThat(result.getNestedGetter().foo, is("BBB"));
    }

    public static final class ResultClass {
        public final NestedClass nestedField = new NestedClass();
        private final NestedClass nestedGetterBackingField = new NestedClass();

        public NestedClass getNestedGetter() {
            return nestedGetterBackingField;
        }
    }

    public static final class NestedClass {
        public String foo;
    }
}

File: dalesbred/src/test/kotlin/org/dalesbred/DatabasePropertyPathBindingTest.kt
package org.dalesbred

import org.dalesbred.testutils.DatabaseProvider.POSTGRESQL
import org.dalesbred.testutils.DatabaseTest
import org.dalesbred.testutils.transactionalTest
import kotlin.test.Test
import kotlin.test.assertEquals

@DatabaseTest(POSTGRESQL)
class DatabasePropertyPathBindingTest(private val db: Database) {

    @Test
    fun `binding to nested paths`() = transactionalTest(db) {
        val result = db.findUnique(ResultClass::class.java, """
                select 'AAA' as "nestedField.foo",
                       'BBB' as "nestedGetter.foo"
                    from (VALUES (0))""")

        assertEquals("AAA", result.nestedField.foo)
        assertEquals("BBB", result.nestedGetter.foo)
    }

    class ResultClass {
        val nestedField = NestedClass()
        val nestedGetter = NestedClass()
    }

    class NestedClass {
        lateinit var foo: String
    }
}
